src/tracer.c

```diff
@@ -127,61 +186,38 @@ handle_syscall(pid_t pid, const struct ptrace_syscall_info *entry,
 	return;			       // return on syscall failure
     }
 
-    const int       syscall = entry->entry.nr;
+    const int syscall = entry->entry.nr;
+    const int fd = exit->exit.rval;
 
     if (syscall != SYS_open && syscall != SYS_creat && syscall != SYS_openat
 	&& syscall != SYS_close) {
 	return;			       // return if we don't care about
 	// tracking said syscall
     }
 
-    PROCESS_INFO   *pinfo = find(pid);
-
     if (syscall == SYS_close) {
-	const int       fd = entry->entry.args[0];
-	FILE_INFO      *f = pinfo->finfo + pinfo->open_files[fd];
-
-	f->hash = hash_file(f->path);
+	handle_close(entry->entry.args[0],
+		     pinfo->finfo + pinfo->open_files[fd]);
 	return;
     }
 
-    FILE_INFO      *finfo = next_finfo(pinfo);
-
-    const int       fd = exit->exit.rval;
-
-    if (fd >= 1024)		       // more than 1024 files open
-	// concurrently
-    {
-	error(EXIT_FAILURE, errno,
-	      "limit of 1024 open files exceeded for process %d", pid);
-    }
-
-    pinfo->open_files[fd] = pinfo->numfinfo;
-
-    if (syscall == SYS_open || syscall == SYS_creat) {
-	finfo->path = get_str_from_process(pid, (void *) entry->entry.args[0]);
-	finfo->purpose = entry->entry.args[1];
-    } else if (syscall == SYS_openat) {
-	finfo->path = get_str_from_process(pid, (void *) entry->entry.args[1]);
-	finfo->purpose = entry->entry.args[2];
-    }
+    handle_open(pid, fd, syscall, entry->entry.args);
```

Comment:
Since everything has moved out, add this in an `if` (case for `open`, `creat` and `openat`)

Reply:
@zvr It's kind of assumed since we check if the syscall is `open` | `creat` | `openat` | `close` at the beginning of the function. But it will be done anyways since we will separate `openat` from `handle_open()`.